### PR TITLE
Feat: remove quant metadata from quantlayer

### PR DESCRIPTION
**Pull request by Giuseppe5**
Quant layer should not expose any attribute of their underlying proxys (e.g, quant_output_scale in QuantLayer).

Similarly, caching should be delegated to Proxy.

Missing tasks:

- [x] Remove comments
- [x] Armonize interface (`is_signed` vs `signed()`)
- [x] Move caching all the way down to proxy (including flags to enable/disable caching)

As task for another PR, we should programmatically generate all the methods that were removed in this PR, such as layer.weight_quant_scale(). The Proxy should expose which methods the layer can pick up, and then generate that using proxy suffix + name of the method.

### CODE REVIEW

File: src/brevitas/proxy/parameter_quant.py [hunk 1]
```diff
@@ -126,6 +139,11 @@ def forward(self, x: torch.Tensor) -> Union[Tensor, QuantTensor]:
 
 class DecoupledWeightQuantWithInputProxyFromInjector(DecoupledWeightQuantProxyFromInjector):
 
+    def __init__(self, quant_layer: nn.Module, quant_injector: Injector) -> None:
+        super().__init__(quant_layer, quant_injector)
+        # Necessary for export
+        self._cached_act = None
```
Comment (Giuseppe5) [inline src/brevitas/proxy/parameter_quant.py:145]: Remove

File: src/brevitas/nn/mixin/act.py [hunk 2]
```diff
@@ -42,22 +40,10 @@ def is_input_quant_enabled(self):
     def is_quant_input_narrow_range(self):  # TODO make abstract once narrow range can be cached
         return self.input_quant.is_narrow_range
```
Comment (Giuseppe5) [inline src/brevitas/nn/mixin/act.py:41]: Remove this and quant_enabled too

File: src/brevitas/nn/mixin/parameter.py [hunk 3]
```diff
@@ -49,9 +51,9 @@ def is_weight_quant_enabled(self):
     def is_quant_weight_narrow_range(self):
```
Comment (Giuseppe5) [inline src/brevitas/nn/mixin/parameter.py:51]: remove this and quant_enabled

File: src/brevitas/nn/quant_layer.py [hunk 4]
```diff
@@ -70,67 +51,6 @@ def is_output_quant_enabled(self):
     def is_quant_output_narrow_range(self):
```
Comment (Giuseppe5) [inline src/brevitas/nn/quant_layer.py:51]: Remove this and output_quant_enabled

File: src/brevitas/proxy/parameter_quant.py [hunk 5]
```diff
@@ -168,21 +190,46 @@ def requires_input_scale(self) -> bool:
...
+                    "No quant bias cache found, set cache_inference_quant_bias=True and run an "
+                    "inference pass first")
+                return None
+            if self.training:
+                warn("Cached quant bias scale is being used in training mode.")
+            return self._cached_bias.bit_width
         zhs = self._zero_hw_sentinel()
-        bit_width = self.__call__(self.tracked_parameter_list[0], zhs).bit_width
-        return bit_width
+        out = self.__call__(self.tracked_parameter_list[0], zhs)
```
Comment (Giuseppe5) [inline src/brevitas/proxy/parameter_quant.py:231]:
```suggestion
        bit_width = self.__call__(self.tracked_parameter_list[0], zhs).bit_width
        return bit_width
```

File: src/brevitas/export/common/handler/base.py [hunk 6]
```diff
@@ -130,21 +131,25 @@ def zero_point_with_dtype(cls, signed, bit_width, zero_point):
...
+        bit_width = module.input_quant.bit_width()
         return cls.zero_point_with_dtype(signed, bit_width, zero_point)
 
     @classmethod
     def quant_weight_zero_point(cls, module):
-        signed = module.is_quant_weight_signed
-        zero_point = module.quant_weight_zero_point()
-        bit_width = module.quant_weight_bit_width()
+        signed = module.weight_quant.is_signed
+        zero_point = module.quant_weight().zero_point
```
Comment (Giuseppe5) [inline src/brevitas/export/common/handler/base.py:142]: weight_quant.zero_point()

File: src/brevitas/nn/quant_layer.py [hunk 7]
```diff
@@ -42,94 +41,15 @@ def channelwise_separable(self) -> bool:
...
-        return self.quant_act_zero_point()
-
-    def quant_input_bit_width(self):
-        if self.is_input_quant_enabled:
-            return self.input_quant.bit_width()
-        elif self._cached_inp is not None:
-            return self._cached_inp.bit_width
-        else:
-            return None
+    # @property
```
Comment (Giuseppe5) [inline src/brevitas/nn/quant_layer.py:46]: remove

File: src/brevitas/export/torch/qoperator/handler/parameter.py [hunk 8]
```diff
@@ -48,19 +48,19 @@ def prepare_bias(cls, module: QuantWBIOL):
...
-        cls.validate_8b_bit_width(module.quant_input_bit_width(), le_then=False)
-        cls.validate_8b_bit_width(module.quant_output_bit_width(), le_then=False)
-        scale = module.quant_weight_scale()
+        cls.validate_bit_width(module.weight_quant.bit_width(), 7, le_then=True)
+        cls.validate_8b_bit_width(module.input_quant.bit_width(), le_then=False)
+        cls.validate_8b_bit_width(module.output_quant.bit_width(), le_then=False)
+        scale = module.weight_quant.scale()
         zero_point = cls.quant_weight_zero_point(module)
-        signed = module.is_quant_weight_signed
+        signed = module.weight_quant.is_signed
```
Comment (Giuseppe5) [inline src/brevitas/export/torch/qoperator/handler/parameter.py:56]: Signed()

File: src/brevitas_examples/super_resolution/utils/evaluate.py [hunk 9]
```diff
@@ -56,7 +56,7 @@ def _calc_min_acc_bit_width(module: QuantWBIOL) -> Tensor:
 
     # bit-width and sign need to come from the quant tensor of the preceding layer if no io_quant
     input_bit_width = module.quant_input_bit_width()
```
Comment (Giuseppe5) [inline src/brevitas_examples/super_resolution/utils/evaluate.py:58]: change and check the rest of the examples

**Comment by Giuseppe5** [comment]
Remove int_weights

**Review by costigt-dev — APPROVED** [review]
looks good, the changes like quant_weight_scale to quant_weight.scale are nice

**Review by nickfraser — APPROVED** [review]
LGTM!

File: src/brevitas/export/common/handler/base.py [hunk 10]
```diff
@@ -130,21 +131,25 @@ def zero_point_with_dtype(cls, signed, bit_width, zero_point):
...
+        zero_point = module.weight_quant.zero_point()
+        bit_width = module.weight_quant.bit_width()
         return cls.zero_point_with_dtype(signed, bit_width, zero_point)
 
     @classmethod
     def quant_output_zero_point(cls, module):
-        signed = module.is_quant_output_signed
-        zero_point = module.quant_output_zero_point()
-        bit_width = module.quant_output_bit_width()
+        if isinstance(module, QuantNonLinearActLayer):
```
Comment (Giuseppe5) [inline src/brevitas/export/common/handler/base.py:148]: Check whether these functions are still used

**Pull request merged**